File: cell.py
import random
from colours import *

class Cell:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.alive      = True
        self.incubation = -1             # Days of incubation before the cell is infectious
        self.infected   = False          # If the cell is infected or not
        self.duration   = -1             # How many days to finish and recover or to get worse
        self.inmunity   = -1             # How inmune the cell is to infection after recovery or not
        self.medication = False          # If the cell has taken its medication or not
        self.quarantined = False         # If the cell is in quarantine
        self.color      = WHITE
        self.quaratineDays = -1 

    def __str__(self):
        out = "Cell " + str(self.x) + ", " + str(self.y) + " is: "
        if self.alive:
            if self.infected:
                out = out + "Infected" 
                if self.incubation > -1:
                    out = out + "\n\tIncubation days remaining: " + str(self.incubation) + "\n\tThe days to finish or recover are " + str(self.duration)
                if self.quarantined:
                    out = out + "\n\tThe cell is in quarantine" 
                if self.medication:
                    out = out + "\n\tThe cell has take a medication" 
                   
            else:
                if self.color == WHITE:
                    out = out + "Not Infected yet"
                if self.color == GREEN:
                    out = out + "Infected but recover \n\t The inmunity is " + str(self.inmunity)
        else:
            out = out + "Die"
        return out

    # Cell become infected  
    def infect(self, incubation, duration):
        self.infected   = True
        self.color      = YELLOW
        self.incubation = incubation
        self.duration   = duration

    # Cell become recovered
    def recover(self, inmunity):
        # print("me he recuperado")
        self.infected = False
        self.incubation = -1
        self.duration   = -1
        self.inmunity   = inmunity
        self.color      = GREEN

    def die(self):
        self.alive    = False
        self.infected = False
        self.duration   = -1
        self.color = BLACK

    def quarantine(self):
        self.quarantined = True
        self.color = BLUE
        self.quaratineDays = 10

    def medicate(self, medEfficacy):
        if random.random() < medEfficacy:
            return True # The medication works and the cell recover
        else:
            self.medication = True
            return False # The medication doesn't work and the cell stay infected

    def process(self, deadliness):
        if self.infected:
            if self.incubation > 0:
                self.incubation = self.incubation - 1
                return 0 # The cell stay incubating
            else:
                self.color = RED   # Red Ya ha incubado el virus por lo que ya es infeccioso
                if self.duration > 0:
                    self.duration = self.duration - 1
                    return 3 # The cell stay infected
                else: # If the virus has been incubated and the duration of the disease ends
                    if random.random() > deadliness:
                        return 1 # The cell recovers
                    else:
                        return 2 # The cell dies



    def processQuarantine(self, quarentEfficacy):
        if self.quaratineDays > 0:
            self.quaratineDays = self.quaratineDays -1
            return 3  # The cell stay in quarantine
        else:
            self.quaratineDays = -1
            self.quarantine = False
            if random.random() < quarentEfficacy:
                return 1 # The cell recovers
            else:
                self.color = RED
                return 2 # The cell stay infected
            

File: compiler/cell.cpp
#include <stdlib.h>
#include <time.h>
#include "cell.h"

// class Cell{
//    public:
//     bool alive;
//     int incubation;           //Days of incubation before the cell is infectious
//     bool infected;            // If the cell is infected or not
//     int duration;             // How many days to finish and recover or to get worse
//     int inmunity;             // How inmune the cell is to infection after recovery or not
//     bool medication;          // If the cell has taken its medication or not
//     bool quarantined;         // If the cell is in quarantine
//     int state;
//     int quaratineDays; 

//     Cell();
//     void infect(int, int);

// };


// Cell::Cell (){
//     alive         = true;
//     state         = 0;
//     infected      = false;       // If the cell is infected or not

//     //Parámetros a partir de aquí
//     incubation    = -1;          // Days of incubation before the cell is infectious
//     duration      = -1;          // How many days to finish and recover or to get worse
//     inmunity      = -1;          // How inmune the cell is to infection after recovery or not
//     medication    = false;       // If the cell has taken its medication or not
//     quarantined   = false;       // If the cell is in quarantine
//     quaratineDays = -1;
// }

// //Cell become infected  
// void Cell::infect(int incubation, int duration){
//     infected   = true;
//     state      = 1;
//     incubation = incubation;
//     duration   = duration;
// }

// // Cell become recovered
// void recover(int inmunity){
//     // print("me he recuperado")
//     infected = false;
//     incubation = -1;
//     duration   = -1;
//     inmunity   = inmunity;
//     state      = 3;
// }

// void die(){
//     alive    = false;
//     infected = false;
//     duration = -1;
//     state    = 5;
// }

// void quarantine(){
//     quarantined   = true;
//     state         = 4;
//     quaratineDays = 10;
// }

// // bool medicate(double medEfficacy){
// //     srand(time(NULL));
// //     double num = (rand() % (1001))/1000.0;
// //     if (num < medEfficacy) {
// //         return true; // The medication works and the cell recover
// //     } else {
// //         medication = true;
// //         return false; // The medication doesn't work and the cell stay infected
// //     }
// // }

// int process(double deadliness){
//     if(infected){
//         if (incubation > 0) {
//             incubation = incubation - 1;
//             return 0; // The cell stay incubating
//         } else {
//             state = 2;  // Red Ya ha incubado el virus por lo que ya es infeccioso
//             if (duration > 0) {
//                 duration = duration - 1;
//                 return 3; // The cell stay infected
//             } else { // If the virus has been incubated and the duration of the disease ends
//                 srand(time(NULL));
//                 double num = (rand() % (1001))/1000.0;
//                 if (num > deadliness) {
//                     return 1; // The cell recovers
//                 } else {
//                     return 2; // The cell dies
//                 }
//             }
//         }
//     }
//     return -1;
// }



// int processQuarantine(double quarentEfficacy){
//     if (quaratineDays > 0) {
//         quaratineDays = quaratineDays - 1;
//         return 3;  // The cell stay in quarantine
//     }else{
//         quaratineDays = -1;
//         quarantined = false;
//         srand(time(NULL));
//         double num = (rand() % (1001))/1000.0;
//         if (num < quarentEfficacy) {
//             return 1; // The cell recovers
//         }else{
//             if(incubation > 0) { state = 1;}
//             else { state = 2; }
//             return 2; // The cell stay infected
//         }
//     }
// }

